### PR TITLE
Additional routes for getting payments

### PR DESCRIPTION
**Pull request by oxifinch**
Added new routes for fetching payments.
- `/payments/user?id` to get all payments belonging to a User
- `/payments/budgetcategory?id` to get all payments in a specific BudgetCategory
- `/payments/category?id` to get all payments from a User's Category across all their Budgets

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -38,9 +38,9 @@ func (db *Database) Seed() {
 	db.Exec("DELETE FROM payments")
 
 	newUsers := []User{
-		{Username: "joseph", Password: "secret01"},
-		{Username: "jean-paul", Password: "secret02"},
-		{Username: "bubby", Password: "secret03"},
+		{Username: "joseph", Password: "secret01", Currency: "SEK"},
```
Comment (fholmqvist) [inline database/database.go]: Time to make some types!

```
type Currency uint

const (
    USD Currency = iota
    SEK
    ...
)
```

File: database/models.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ type Category struct {
 	Description string
 	Color       string
 	UserID      uint
+	User        User
```
Comment (fholmqvist) [inline database/models.go:39]: We're starting to get into situations where every model embeds every other model, resulting in giant (picture the JSON) structs with dependencies all over the place.

Next PR will need to clean these up, we only reference the `ID` for the needed resource and our endpoints will have custom result types as listed in:
```
https://gorm.io/docs/query.html#Joins
```

`user` can have `categories`, sure, but the inverse is bonkers.

File: payments/repo.go [hunk 3]
```diff
@@ -35,14 +35,64 @@ func (p *PaymentRepo) GetAllByBudgetID(id uint) ([]database.Payment, error) {
...
 		Order("date desc").
 		Find(&payments).Error
 
 	return payments, err
 }
 
+func (p *PaymentRepo) GetAllByCategoryID(id uint) ([]database.Payment, error) {
+	var payments []database.Payment
+
+	// TODO: Couldn't figure out how to properly filter by Category.ID, redo
```
Comment (fholmqvist) [inline payments/repo.go]: ```.Where("something.field = ?", id)```

File: router/payments.go [hunk 4]
```diff
@@ -37,6 +43,96 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
+		displayErrorPage(w, r, http.StatusBadRequest,
+			"The request included an invalid resource ID. Check the URL and try again.")
+	}
+
+	ps, err := rt.PaymentRepo.GetAllByBudgetCategoryID(uint(id))
+	if err != nil {
+		log.Fatalf("handlePaymentsBudgetCategory: %v\n", err)
+	}
+
+	data := struct {
```
Comment (fholmqvist) [inline router/payments.go]: No need to wrap a struct into a new struct with a single field.

File: router/payments.go [hunk 5]
```diff
@@ -37,6 +43,96 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
+		displayErrorPage(w, r, http.StatusBadRequest,
+			"The request included an invalid resource ID. Check the URL and try again.")
+	}
+
+	ps, err := rt.PaymentRepo.GetAllByCategoryID(uint(id))
+	if err != nil {
+		log.Fatalf("handlePaymentsCategory: %v\n", err)
+	}
+
+	data := struct {
```
Comment (fholmqvist) [inline router/payments.go]: Same.

File: payments/repo.go [hunk 6]
```diff
@@ -35,14 +35,64 @@ func (p *PaymentRepo) GetAllByBudgetID(id uint) ([]database.Payment, error) {
...
+		}
+	}
+
+	return filteredPayments, err
+}
+
+func (p *PaymentRepo) GetAllByUserID(id uint) ([]database.Payment, error) {
+	var payments []database.Payment
+
+	// TODO: How do I filter out by BudgetCategory.Category.UserID = id in the query?
```
Comment (fholmqvist) [inline payments/repo.go]: Same.

File: router/payments.go [hunk 7]
```diff
@@ -37,6 +43,96 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
+		displayErrorPage(w, r, http.StatusBadRequest,
+			"The request included an invalid resource ID. Check the URL and try again.")
+	}
+
+	ps, err := rt.PaymentRepo.GetAllByUserID(uint(id))
+	if err != nil {
+		log.Fatalf("handlePaymentsUser: %v\n", err)
+	}
+
+	data := struct {
```
Comment (fholmqvist) [inline router/payments.go]: Same.

**Review by fholmqvist — COMMENTED** [review]
A bit more this time.

File: payments/repo.go (hunk 3 above)
Comment (oxifinch) [reply payments/repo.go]: categories.user_id is not a column, unless we join the table. And I have no idea how to use GORM Joins yet :D

File: router/payments.go [hunk 8]
```diff
@@ -19,20 +24,86 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
 			"The request included an invalid resource ID. Check the URL and try again.")
 	}
 
-	ps, err := rt.PaymentRepo.GetAllByBudgetID(uint(id))
+	Payments, err := rt.PaymentRepo.GetAllByBudgetID(uint(id))
```
Comment (fholmqvist) [inline router/payments.go]: Grr!

File: router/payments.go [hunk 9]
```diff
@@ -19,20 +24,86 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
-		log.Fatalf("handlePaymentsBudget: %v\n", err)
+		log.Fatalf("handlePaymentsBudgetCategory: %v\n", err)
+	}
+
+	if id < 1 {
+		displayErrorPage(w, r, http.StatusBadRequest,
+			"The request included an invalid resource ID. Check the URL and try again.")
+	}
+
+	Payments, err := rt.PaymentRepo.GetAllByBudgetCategoryID(uint(id))
```
Comment (fholmqvist) [inline router/payments.go]: Grr!

File: router/payments.go [hunk 10]
```diff
@@ -19,20 +24,86 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		log.Fatalf("handlePaymentsCategory: %v\n", err)
+	}
+
+	if id < 1 {
+		displayErrorPage(w, r, http.StatusBadRequest,
+			"The request included an invalid resource ID. Check the URL and try again.")
+	}
+
+	Payments, err := rt.PaymentRepo.GetAllByCategoryID(uint(id))
```
Comment (fholmqvist) [inline router/payments.go]: Grr!

File: router/payments.go [hunk 11]
```diff
@@ -19,20 +24,86 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
+	// if err != nil {
```
Comment (fholmqvist) [inline router/payments.go:76]: Grrrrrrrrrr!!1

File: router/payments.go [hunk 12]
```diff
@@ -19,20 +24,86 @@ func (rt *Router) handlePaymentsBudget(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		log.Fatalf("handlePaymentsUser: %v\n", err)
+	}
+
+	if id < 1 {
+		displayErrorPage(w, r, http.StatusBadRequest,
+			"The request included an invalid resource ID. Check the URL and try again.")
+	}
+
+	Payments, err := rt.PaymentRepo.GetAllByUserID(uint(id))
```
Comment (fholmqvist) [inline router/payments.go]: Grr!

**Review by fholmqvist — COMMENTED** [review]
Grr!

**Review by fholmqvist — APPROVED** [review]

**Pull request merged**